tools/cmd/crd-doc-gen/main.go

```diff
@@ -0,0 +1,95 @@
+package main
+
+import (
+	"go/build"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"kubesphere.io/kubesphere/tools/lib"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/go-openapi/spec"
+	s2iv1alpha1 "github.com/kubesphere/s2ioperator/pkg/apis/devops/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	urlruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/kube-openapi/pkg/common"
+	tenantinstall "kubesphere.io/kubesphere/pkg/apis/tenant/crdinstall"
+	tenantv1alpha1 "kubesphere.io/kubesphere/pkg/apis/tenant/v1alpha1"
+)
+
+func main() {
+
+	var (
+		Scheme = runtime.NewScheme()
+		Codecs = serializer.NewCodecFactory(Scheme)
+	)
+	// tmp install s2i api group because
+	// cannot use Scheme (type *"kubesphere.io/kubesphere/vendor/k8s.io/apimachinery/pkg/runtime".Scheme) as type *"github.com/kubesphere/s2ioperator/vendor/k8s.io/apimachinery/pkg/runtime".Scheme in argument to "github.com/kubesphere/s2ioperator/pkg/apis/devops/install".Install
+	urlruntime.Must(s2iv1alpha1.AddToScheme(Scheme))
+	urlruntime.Must(Scheme.SetVersionPriority(s2iv1alpha1.SchemeGroupVersion))
+
+	tenantinstall.Install(Scheme)
+
+	mapper := meta.NewDefaultRESTMapper(nil)
+
+	mapper.AddSpecific(tenantv1alpha1.SchemeGroupVersion.WithKind(tenantv1alpha1.ResourceKindWorkspace),
```

Comment:
what about servicemesh related crds?

Reply:
Currently lacking the isio-related go language openapi structure, I need to submit some code to github/knative/pkg to generate the code.

Reply:
https://github.com/kubesphere/kubesphere/pull/609/files#diff-b4163ba54c93a4371a0a28df03f28b28R94